lib/restforce/concerns/api.rb

```diff
@@ -361,12 +361,17 @@ def upsert(*args)
       # error returned if the external ID provided matches multiple records (in which
       # case the conflicting IDs can be found by looking at the response on the error)
       def upsert!(sobject, field, attrs)
-        external_id = attrs.
-          fetch(attrs.keys.find { |k, v| k.to_s.downcase == field.to_s.downcase }, nil)
-        attrs_without_field = attrs.
-          reject { |k, v| k.to_s.downcase == field.to_s.downcase }
-        response = api_patch "sobjects/#{sobject}/#{field}/#{URI.encode(external_id)}",
-                             attrs_without_field
+        external_id = attrs.delete(field.to_sym)
```

Comment:
> It might be nice to extract this into a private method rather than have this rather verbose piece of code here. I guess it'd be something like `extract_case_insensitive_string_or_symbol_key_from_hash`. What do you think?

If it won't be used elsewhere, my inclination is to leave it in `upsert!` to avoid indirection when debugging and add comment to describe what's going on.  I'm open to whichever you think makes sense long term.  

> Could we otherwise not document that you need to provide symbol keys and a symbol field name? That feels a bit nicer than having to do these comparisons every time.

The flexibility is nice for consumers of the gem to accommodate whatever sources `field` and `attrs` are coming from and these changes reduce the number of comparisons performed from `attrs.length` to a max of 4 if the match is found on the last check.  Changing the current behavior might be a pain for anyone incidentally relying on it.